lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,78 @@
+    }
+    MOBILE_PLATFORM_VERSION = {
+        # version is found as last entry in android user agent after "/"
+        MobileAppVersionConfig.ANDROID: lambda user_agent: user_agent[user_agent.rfind("/") + 1:],
+        # version is found between first occurrence ( and ; i.e. (version;
+        MobileAppVersionConfig.IOS: lambda user_agent: user_agent[user_agent.find("(") + 1:user_agent.find(";")],
+    }
+
+    def form_factor(self, user_agent):
+        """ extracts mobile platform and app version in use """
+        form_factor = None
+        for platform, user_agent_regex in self.MOBILE_PLATFORM_USER_AGENT_REGEX.iteritems():
+            if re.search(user_agent_regex, user_agent):
+                user_app_version = self.MOBILE_PLATFORM_VERSION[platform](user_agent)
+                # if user app version does not match to X.X.X.Y (acceptable format) then let pass the request
+                # as it might belong to older apps, so return None
+                if self.validate_version(user_app_version):
+                    return {"platform": platform, "user_app_version": user_app_version}
+                break
+        return form_factor
+
+    def validate_version(self, user_app_version):
+        """ if user app version confronts to X.X.X.Y format return X.X.X else None"""
+        if re.match(settings.MOBILE_APP_VERSION_REGEX, user_app_version):
+            return True
+
+    def process_request(self, request):
+        """
+        If request is from mobile native app and confront to X.X.X.Y version format (where X is a number and
+        Y is optional alphanumeric), and return Http 426 if version upgrade is required; else let pass the request
+        """
+        if is_request_from_mobile_app(request):
+            user_agent = request.META.get('HTTP_USER_AGENT')
+            form_factor = self.form_factor(user_agent)
+            if form_factor:
+                app_version_config = MobileAppVersionConfig.current(form_factor["platform"])
+                if app_version_config.is_outdated_version(form_factor["user_app_version"]):
+                    return HttpResponse(status=426)
+
+    def process_response(self, _request, response):
+        """
+        If request is from mobile native app and confront to X.X.X.Y version format (where X is a number and
+        Y is optional alphanumeric), then add headers to response;
+        1. EDX-APP-LATEST-VERSION; if user app version < latest available version
+        2. EDX-APP-UPGRADE-DATE; if user app version < next supported version
+        """
+        if is_request_from_mobile_app(_request):
+            user_agent = _request.META.get('HTTP_USER_AGENT')
+            form_factor = self.form_factor(user_agent)
+            if form_factor:
+                user_app_version = form_factor["user_app_version"]
+                app_version_config = MobileAppVersionConfig.current(form_factor["platform"])
```

Comment:
How costly are these computations?  Consider caching the parsed version in the `request` object during `process_request` so you don't need to recompute it here.